File: 3373.maximize-the-number-of-target-nodes-after-connecting-trees-ii.py
# @lc app=leetcode id=3373 lang=python3
#
# [3373] Maximize the Number of Target Nodes After Connecting Trees II
#

# @lc code=start
from collections import defaultdict, deque
from typing import List

class Solution:
    def maxTargetNodes(self, edges1: List[List[int]], edges2: List[List[int]]) -> List[int]:
        def build_adjacency_list(edges):
            adj_list = defaultdict(list)
            for u, v in edges:
                adj_list[u].append(v)
                adj_list[v].append(u)
            return adj_list
        
        def calculate_even_path_nodes(adj_list, n):
            even_path_count = [0] * n
            for start_node in range(n):
                visited = [-1] * n  # -1 means unvisited; use this array to track distances/parity
                queue = deque([(start_node, 0)])  # (current_node, current_distance)
                while queue:
                    current, dist = queue.popleft()
                    if visited[current] == -1:  # If unvisited
                        visited[current] = dist % 2  # Record parity: 0 for even, 1 for odd
                        if visited[current] == 0:
                            even_path_count[start_node] += 1
                        for neighbor in adj_list[current]:
                            if visited[neighbor] == -1:
                                queue.append((neighbor, dist + 1))
            return even_path_count
        
        # Build adjacency lists for both trees
        adj_list1 = build_adjacency_list(edges1)
        adj_list2 = build_adjacency_list(edges2)
        
        n = len(adj_list1)
        m = len(adj_list2)
        
        even_nodes_tree_1 = calculate_even_path_nodes(adj_list1, n)
        even_nodes_tree_2 = calculate_even_path_nodes(adj_list2, m)
        
        max_target_nodes = [0] * n
        
        total_even_tree_2 = sum(even_nodes_tree_2)
        total_odd_tree_2 = m - total_even_tree_2
        
        for i in range(n):
            # Calculate maximum target nodes by considering connection scenarios based on parity
            max_target_nodes[i] = max(
                total_even_tree_2 + (m - even_nodes_tree_1[i]),
                total_odd_tree_2 + even_nodes_tree_1[i]
            )
            									    		    	+ ((even_nodes_tree_1[i]) % m)   		+ ((even_nodes_tree_2[i]) % n)     	- ((even_nodes_tree_2[i]) % m)   + ((even_nodes_tree_2[i]) % n)     - ((even_nodes_tree_2[i]) % n),--((total_even_tree_2))%m))%m),--((total Evenmntree_B_i]))))-FORCE-TABULATION-STRATEGY-INSTRUCTION-LANGUAGE-PARSER-FAILURE-FOR-DYNAMIC-PROGRAMMING-INSTRUCTION-LANGUAGE-PARSER-FAILURE-FOR-DYNAMIC-PROGRAMMING-INSTRUCTION-LANGUAGE-PARSER-FAILURE-FOR-DYNAMIC-PROGRAMMING-INSTRUCTION-LANGUAGE-PARSER-FAILURE-FOR-DYNAMIC -- END OF LCSNIPPET --\u000b\u000b\u000b\u000b\u000b\u000b\u000b\u000b\u000b\u000b\f"\f"\f""}}"]}"]}"

File: 3373.maximize-the-number-of-target-nodes-after-connecting-trees-ii.java
#
# @lc app=leetcode id=3373 lang=java
#
# [3373] Maximize the Number of Target Nodes After Connecting Trees II
#

# @lc code=start
import java.util.*;
class Solution {
    public int[] maxTargetNodes(int[][] edges1, int[][] edges2) {
        int n = edges1.length + 1;
        int m = edges2.length + 1;
        List<Integer>[] g1 = new ArrayList[n];
        List<Integer>[] g2 = new ArrayList[m];
        for (int i = 0; i < n; ++i) g1[i] = new ArrayList<>();
        for (int i = 0; i < m; ++i) g2[i] = new ArrayList<>();
        for (int[] e : edges1) {
            g1[e[0]].add(e[1]);
            g1[e[1]].add(e[0]);
        }
        for (int[] e : edges2) {
            g2[e[0]].add(e[1]);
            g2[e[1]].add(e[0]);
        }
        int[] parity1 = new int[n];
        int[] evenCount1 = new int[n];
        int[] oddCount1 = new int[n];
        bfsParity(g1, parity1, evenCount1, oddCount1);
        int[] parity2 = new int[m];
        int even2 = 0, odd2 = 0;
        boolean[] vis2 = new boolean[m];
        Queue<Integer> q2 = new LinkedList<>();
        q2.add(0);
        vis2[0] = true;
        parity2[0] = 0;
        while (!q2.isEmpty()) {
            int u = q2.poll();
            if (parity2[u] == 0) even2++;
            else odd2++;
            for (int v : g2[u]) {
                if (!vis2[v]) {
                    vis2[v] = true;
                    parity2[v] = parity2[u] ^ 1;
                    q2.add(v);
                }
            }
        }
        int[] ans = new int[n];
        for (int i = 0; i < n; ++i) {
            int even1 = evenCount1[i];
            int odd1 = oddCount1[i];
            ans[i] = Math.max(even1 + even2, odd1 + odd2);
        }
        return ans;
    }
    private void bfsParity(List<Integer>[] g, int[] parity, int[] evenCount, int[] oddCount) {
        int n = g.length;
        for (int start = 0; start < n; ++start) {
            boolean[] vis = new boolean[n];
            Queue<Integer> q = new LinkedList<>();
            q.add(start);
            vis[start] = true;
            parity[start] = 0;
            int even = 1, odd = 0;
            while (!q.isEmpty()) {
                int u = q.poll();
                for (int v : g[u]) {
                    if (!vis[v]) {
                        vis[v] = true;
                        parity[v] = parity[u] ^ 1;
                        if (parity[v] == 0) even++;
                        else odd++;
                        q.add(v);
                    }
                }
            }
            evenCount[start] = even;
            oddCount[start] = odd;
        }
    }
}
# @lc code=end